File: Mediconnet-Frontend/src/app/services/assurance.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';

// ==================== INTERFACES ====================

export interface AssuranceListItem {
  idAssurance: number;
  nom: string;
  typeAssurance: string;
  groupe?: string;
  zoneCouverture?: string;
  isActive: boolean;
  nombrePatientsAssures: number;
}

export interface AssuranceDetail {
  idAssurance: number;
  nom: string;
  typeAssurance: string;
  siteWeb?: string;
  telephoneServiceClient?: string;
  groupe?: string;
  paysOrigine?: string;
  statutJuridique?: string;
  description?: string;
  typeCouverture?: string;
  isComplementaire: boolean;
  categorieBeneficiaires?: string;
  conditionsAdhesion?: string;
  zoneCouverture?: string;
  modePaiement?: string;
  isActive: boolean;
  createdAt: string;
  updatedAt: string;
  nombrePatientsAssures: number;
}

export interface CreateAssuranceDto {
  nom: string;
  typeAssurance: string;
  siteWeb?: string;
  telephoneServiceClient?: string;
  groupe?: string;
  paysOrigine?: string;
  statutJuridique?: string;
  description?: string;
  typeCouverture?: string;
  isComplementaire?: boolean;
  categorieBeneficiaires?: string;
  conditionsAdhesion?: string;
  zoneCouverture?: string;
  modePaiement?: string;
  isActive?: boolean;
}

export interface PatientAssuranceInfo {
  estAssure: boolean;
  assuranceId?: number;
  nomAssurance?: string;
  typeAssurance?: string;
  couvertureAssurance?: number; // Taux de couverture du patient
  numeroCarteAssurance?: string;
  dateDebutValidite?: string;
  dateFinValidite?: string;
  estValide: boolean;
  tauxEffectif?: number; // Taux effectif calculé
}

export interface UpdatePatientAssuranceDto {
  assuranceId?: number;
  numeroCarteAssurance?: string;
  dateDebutValidite?: string;
  dateFinValidite?: string;
  couvertureAssurance?: number;
}

export interface AssuranceListResponse {
  success: boolean;
  data: AssuranceListItem[];
  total: number;
  totalActives: number;
}

export interface AssuranceResponse {
  success: boolean;
  message: string;
  data?: AssuranceDetail;
}

export interface PatientAssuranceResponse {
  success: boolean;
  message: string;
  data?: PatientAssuranceInfo;
}

export interface AssuranceFilter {
  typeAssurance?: string;
  zoneCouverture?: string;
  isActive?: boolean;
  recherche?: string;
  page?: number;
  pageSize?: number;
}

// ==================== CONSTANTES ====================

export const TYPES_ASSURANCE = [
  { value: 'privee', label: 'Privée' },
  { value: 'publique', label: 'Publique' },
  { value: 'mutuelle', label: 'Mutuelle' },
  { value: 'micro_assurance', label: 'Micro-assurance' },
  { value: 'programme_public', label: 'Programme public' }
];

export const STATUTS_JURIDIQUES = [
  { value: 'compagnie', label: 'Compagnie' },
  { value: 'mutuelle', label: 'Mutuelle' },
  { value: 'organisme_public', label: 'Organisme public' },
  { value: 'cooperative', label: 'Coopérative' }
];

export const ZONES_COUVERTURE = [
  { value: 'national', label: 'National' },
  { value: 'regional', label: 'Régional' },
  { value: 'rural', label: 'Rural' },
  { value: 'diaspora', label: 'Diaspora' },
  { value: 'international', label: 'International' }
];

export const MODES_PAIEMENT = [
  { value: 'mobile_money', label: 'Mobile Money' },
  { value: 'entreprise', label: 'Entreprise' },
  { value: 'cotisations', label: 'Cotisations' },
  { value: 'prelevement', label: 'Prélèvement bancaire' },
  { value: 'carte', label: 'Carte bancaire' }
];

// ==================== SERVICE ====================

@Injectable({
  providedIn: 'root'
})
export class AssuranceService {
  private apiUrl = `${environment.apiUrl}/assurance`;

  constructor(private http: HttpClient) {}

  // ==================== ASSURANCES ====================

  /**
   * Récupérer la liste des assurances avec filtres
   */
  getAssurances(filter?: AssuranceFilter): Observable<AssuranceListResponse> {
    let params = new HttpParams();
    if (filter) {
      if (filter.typeAssurance) params = params.set('typeAssurance', filter.typeAssurance);
      if (filter.zoneCouverture) params = params.set('zoneCouverture', filter.zoneCouverture);
      if (filter.isActive !== undefined) params = params.set('isActive', filter.isActive.toString());
      if (filter.recherche) params = params.set('recherche', filter.recherche);
      if (filter.page) params = params.set('page', filter.page.toString());
      if (filter.pageSize) params = params.set('pageSize', filter.pageSize.toString());
    }
    return this.http.get<AssuranceListResponse>(this.apiUrl, { params });
  }

  /**
   * Récupérer les assurances actives (pour listes déroulantes)
   */
  getAssurancesActives(): Observable<AssuranceListItem[]> {
    return this.http.get<AssuranceListItem[]>(`${this.apiUrl}/actives`);
  }

  /**
   * Récupérer une assurance par son ID
   */
  getAssuranceById(id: number): Observable<AssuranceDetail> {
    return this.http.get<AssuranceDetail>(`${this.apiUrl}/${id}`);
  }

  /**
   * Créer une nouvelle assurance
   */
  createAssurance(data: CreateAssuranceDto): Observable<AssuranceResponse> {
    return this.http.post<AssuranceResponse>(this.apiUrl, data);
  }

  /**
   * Mettre à jour une assurance
   */
  updateAssurance(id: number, data: CreateAssuranceDto): Observable<AssuranceResponse> {
    return this.http.put<AssuranceResponse>(`${this.apiUrl}/${id}`, data);
  }

  /**
   * Activer/Désactiver une assurance
   */
  toggleAssuranceStatus(id: number): Observable<AssuranceResponse> {
    return this.http.patch<AssuranceResponse>(`${this.apiUrl}/${id}/toggle-status`, {});
  }

  /**
   * Supprimer une assurance
   */
  deleteAssurance(id: number): Observable<AssuranceResponse> {
    return this.http.delete<AssuranceResponse>(`${this.apiUrl}/${id}`);
  }

  // ==================== PATIENT ASSURANCE ====================

  /**
   * Récupérer l'assurance d'un patient
   */
  getPatientAssurance(idPatient: number): Observable<PatientAssuranceInfo> {
    return this.http.get<PatientAssuranceInfo>(`${this.apiUrl}/patient/${idPatient}`);
  }

  /**
   * Mettre à jour l'assurance d'un patient
   */
  updatePatientAssurance(idPatient: number, data: UpdatePatientAssuranceDto): Observable<PatientAssuranceResponse> {
    return this.http.put<PatientAssuranceResponse>(`${this.apiUrl}/patient/${idPatient}`, data);
  }

  /**
   * Retirer l'assurance d'un patient
   */
  removePatientAssurance(idPatient: number): Observable<PatientAssuranceResponse> {
    return this.http.delete<PatientAssuranceResponse>(`${this.apiUrl}/patient/${idPatient}`);
  }
}

File: Mediconnet-Backend/Services/AssuranceService.cs
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Logging;
using Mediconnet_Backend.Core.Entities;
using Mediconnet_Backend.Core.Interfaces.Services;
using Mediconnet_Backend.Data;
using Mediconnet_Backend.DTOs.Assurance;

namespace Mediconnet_Backend.Services;

public class AssuranceService : IAssuranceService
{
    private readonly ApplicationDbContext _context;
    private readonly ILogger<AssuranceService> _logger;

    public AssuranceService(ApplicationDbContext context, ILogger<AssuranceService> logger)
    {
        _context = context;
        _logger = logger;
    }

    // ==================== ASSURANCES ====================

    public async Task<AssuranceListResponse> GetAssurancesAsync(AssuranceFilterDto? filter = null)
    {
        var query = _context.Assurances.Include(a => a.Patients).AsQueryable();

        if (filter != null)
        {
            if (!string.IsNullOrEmpty(filter.TypeAssurance))
                query = query.Where(a => a.TypeAssurance == filter.TypeAssurance);

            if (!string.IsNullOrEmpty(filter.ZoneCouverture))
                query = query.Where(a => a.ZoneCouverture == filter.ZoneCouverture);

            if (filter.IsActive.HasValue)
                query = query.Where(a => a.IsActive == filter.IsActive.Value);

            if (!string.IsNullOrEmpty(filter.Recherche))
            {
                var search = filter.Recherche.ToLower();
                query = query.Where(a => 
                    a.Nom.ToLower().Contains(search) ||
                    (a.Groupe != null && a.Groupe.ToLower().Contains(search)) ||
                    (a.Description != null && a.Description.ToLower().Contains(search)));
            }
        }

        var total = await query.CountAsync();
        var totalActives = await query.CountAsync(a => a.IsActive);

        var page = filter?.Page ?? 1;
        var pageSize = filter?.PageSize ?? 20;
        
        var assurances = await query
            .OrderBy(a => a.Nom)
            .Skip((page - 1) * pageSize)
            .Take(pageSize)
            .Select(a => new AssuranceListDto
            {
                IdAssurance = a.IdAssurance,
                Nom = a.Nom,
                TypeAssurance = a.TypeAssurance,
                Groupe = a.Groupe,
                ZoneCouverture = a.ZoneCouverture,
                IsActive = a.IsActive,
                NombrePatientsAssures = a.Patients.Count
            })
            .ToListAsync();

        return new AssuranceListResponse
        {
            Success = true,
            Data = assurances,
            Total = total,
            TotalActives = totalActives
        };
    }

    public async Task<List<AssuranceListDto>> GetAssurancesActivesAsync()
    {
        return await _context.Assurances
            .Where(a => a.IsActive)
            .OrderBy(a => a.Nom)
            .Select(a => new AssuranceListDto
            {
                IdAssurance = a.IdAssurance,
                Nom = a.Nom,
                TypeAssurance = a.TypeAssurance,
                Groupe = a.Groupe,
                ZoneCouverture = a.ZoneCouverture,
                IsActive = a.IsActive,
                NombrePatientsAssures = 0
            })
            .ToListAsync();
    }

    public async Task<AssuranceDetailDto?> GetAssuranceByIdAsync(int idAssurance)
    {
        var assurance = await _context.Assurances
            .Include(a => a.Patients)
            .FirstOrDefaultAsync(a => a.IdAssurance == idAssurance);

        if (assurance == null) return null;

        return new AssuranceDetailDto
        {
            IdAssurance = assurance.IdAssurance,
            Nom = assurance.Nom,
            TypeAssurance = assurance.TypeAssurance,
            SiteWeb = assurance.SiteWeb,
            TelephoneServiceClient = assurance.TelephoneServiceClient,
            Groupe = assurance.Groupe,
            PaysOrigine = assurance.PaysOrigine,
            StatutJuridique = assurance.StatutJuridique,
            Description = assurance.Description,
            TypeCouverture = assurance.TypeCouverture,
            IsComplementaire = assurance.IsComplementaire,
            CategorieBeneficiaires = assurance.CategorieBeneficiaires,
            ConditionsAdhesion = assurance.ConditionsAdhesion,
            ZoneCouverture = assurance.ZoneCouverture,
            ModePaiement = assurance.ModePaiement,
            IsActive = assurance.IsActive,
            CreatedAt = assurance.CreatedAt,
            UpdatedAt = assurance.UpdatedAt,
            NombrePatientsAssures = assurance.Patients.Count
        };
    }

    public async Task<AssuranceResponse> CreateAssuranceAsync(CreateAssuranceDto dto)
    {
        try
        {
            var exists = await _context.Assurances.AnyAsync(a => a.Nom.ToLower() == dto.Nom.ToLower());
            if (exists)
            {
                return new AssuranceResponse { Success = false, Message = "Une assurance avec ce nom existe déjà" };
            }

            var assurance = new Assurance
            {
                Nom = dto.Nom,
                TypeAssurance = dto.TypeAssurance,
                SiteWeb = dto.SiteWeb,
                TelephoneServiceClient = dto.TelephoneServiceClient,
                Groupe = dto.Groupe,
                PaysOrigine = dto.PaysOrigine,
                StatutJuridique = dto.StatutJuridique,
                Description = dto.Description,
                TypeCouverture = dto.TypeCouverture,
                IsComplementaire = dto.IsComplementaire,
                CategorieBeneficiaires = dto.CategorieBeneficiaires,
                ConditionsAdhesion = dto.ConditionsAdhesion,
                ZoneCouverture = dto.ZoneCouverture,
                ModePaiement = dto.ModePaiement,
                IsActive = dto.IsActive,
                CreatedAt = DateTime.UtcNow,
                UpdatedAt = DateTime.UtcNow
            };

            _context.Assurances.Add(assurance);
            await _context.SaveChangesAsync();

            _logger.LogInformation($"Assurance créée: {assurance.Nom} (ID: {assurance.IdAssurance})");

            return new AssuranceResponse
            {
                Success = true,
                Message = "Assurance créée avec succès",
                Data = await GetAssuranceByIdAsync(assurance.IdAssurance)
            };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Erreur lors de la création de l'assurance");
            return new AssuranceResponse { Success = false, Message = "Erreur lors de la création de l'assurance" };
        }
    }

    public async Task<AssuranceResponse> UpdateAssuranceAsync(int idAssurance, UpdateAssuranceDto dto)
    {
        try
        {
            var assurance = await _context.Assurances.FindAsync(idAssurance);
            if (assurance == null)
            {
                return new AssuranceResponse { Success = false, Message = "Assurance non trouvée" };
            }

            if (assurance.Nom.ToLower() != dto.Nom.ToLower())
            {
                var exists = await _context.Assurances.AnyAsync(a => a.Nom.ToLower() == dto.Nom.ToLower() && a.IdAssurance != idAssurance);
                if (exists)
                {
                    return new AssuranceResponse { Success = false, Message = "Une assurance avec ce nom existe déjà" };
                }
            }

            assurance.Nom = dto.Nom;
            assurance.TypeAssurance = dto.TypeAssurance;
            assurance.SiteWeb = dto.SiteWeb;
            assurance.TelephoneServiceClient = dto.TelephoneServiceClient;
            assurance.Groupe = dto.Groupe;
            assurance.PaysOrigine = dto.PaysOrigine;
            assurance.StatutJuridique = dto.StatutJuridique;
            assurance.Description = dto.Description;
            assurance.TypeCouverture = dto.TypeCouverture;
            assurance.IsComplementaire = dto.IsComplementaire;
            assurance.CategorieBeneficiaires = dto.CategorieBeneficiaires;
            assurance.ConditionsAdhesion = dto.ConditionsAdhesion;
            assurance.ZoneCouverture = dto.ZoneCouverture;
            assurance.ModePaiement = dto.ModePaiement;
            assurance.IsActive = dto.IsActive;
            assurance.UpdatedAt = DateTime.UtcNow;

            await _context.SaveChangesAsync();

            _logger.LogInformation($"Assurance mise à jour: {assurance.Nom} (ID: {assurance.IdAssurance})");

            return new AssuranceResponse
            {
                Success = true,
                Message = "Assurance mise à jour avec succès",
                Data = await GetAssuranceByIdAsync(assurance.IdAssurance)
            };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Erreur lors de la mise à jour de l'assurance");
            return new AssuranceResponse { Success = false, Message = "Erreur lors de la mise à jour de l'assurance" };
        }
    }

    public async Task<AssuranceResponse> ToggleAssuranceStatusAsync(int idAssurance)
    {
        var assurance = await _context.Assurances.FindAsync(idAssurance);
        if (assurance == null)
        {
            return new AssuranceResponse { Success = false, Message = "Assurance non trouvée" };
        }

        assurance.IsActive = !assurance.IsActive;
        assurance.UpdatedAt = DateTime.UtcNow;
        await _context.SaveChangesAsync();

        return new AssuranceResponse
        {
            Success = true,
            Message = $"Assurance {(assurance.IsActive ? "activée" : "désactivée")} avec succès",
            Data = await GetAssuranceByIdAsync(assurance.IdAssurance)
        };
    }

    public async Task<AssuranceResponse> DeleteAssuranceAsync(int idAssurance)
    {
        var assurance = await _context.Assurances
            .Include(a => a.Patients)
            .FirstOrDefaultAsync(a => a.IdAssurance == idAssurance);

        if (assurance == null)
        {
            return new AssuranceResponse { Success = false, Message = "Assurance non trouvée" };
        }

        if (assurance.Patients.Any())
        {
            return new AssuranceResponse
            {
                Success = false,
                Message = "Impossible de supprimer une assurance liée à des patients. Désactivez-la plutôt."
            };
        }

        _context.Assurances.Remove(assurance);
        await _context.SaveChangesAsync();

        _logger.LogInformation($"Assurance supprimée: {assurance.Nom} (ID: {idAssurance})");

        return new AssuranceResponse { Success = true, Message = "Assurance supprimée avec succès" };
    }

    // ==================== PATIENT ASSURANCE ====================

    public async Task<PatientAssuranceInfoDto?> GetPatientAssuranceAsync(int idPatient)
    {
        var patient = await _context.Patients
            .Include(p => p.Assurance)
            .FirstOrDefaultAsync(p => p.IdUser == idPatient);

        if (patient == null) return null;

        var estValide = patient.AssuranceId.HasValue &&
                        (!patient.DateFinValidite.HasValue || patient.DateFinValidite.Value >= DateTime.UtcNow);

        return new PatientAssuranceInfoDto
        {
            EstAssure = patient.AssuranceId.HasValue,
            AssuranceId = patient.AssuranceId,
            NomAssurance = patient.Assurance?.Nom,
            TypeAssurance = patient.Assurance?.TypeAssurance,
            CouvertureAssurance = patient.CouvertureAssurance,
            NumeroCarteAssurance = patient.NumeroCarteAssurance,
            DateDebutValidite = patient.DateDebutValidite,
            DateFinValidite = patient.DateFinValidite,
            EstValide = estValide
        };
    }

    public async Task<PatientAssuranceResponse> UpdatePatientAssuranceAsync(int idPatient, UpdatePatientAssuranceDto dto)
    {
        try
        {
            var patient = await _context.Patients.FindAsync(idPatient);
            if (patient == null)
            {
                return new PatientAssuranceResponse { Success = false, Message = "Patient non trouvé" };
            }

            // Vérifier que l'assurance existe si fournie
            if (dto.AssuranceId.HasValue)
            {
                var assurance = await _context.Assurances.FindAsync(dto.AssuranceId.Value);
                if (assurance == null || !assurance.IsActive)
                {
                    return new PatientAssuranceResponse { Success = false, Message = "Assurance non trouvée ou inactive" };
                }
            }

            patient.AssuranceId = dto.AssuranceId;
            patient.NumeroCarteAssurance = dto.NumeroCarteAssurance;
            patient.DateDebutValidite = dto.DateDebutValidite;
            patient.DateFinValidite = dto.DateFinValidite;
            patient.CouvertureAssurance = dto.CouvertureAssurance;

            await _context.SaveChangesAsync();

            _logger.LogInformation($"Assurance patient mise à jour: Patient {idPatient}, Assurance {dto.AssuranceId}");

            return new PatientAssuranceResponse
            {
                Success = true,
                Message = "Assurance du patient mise à jour avec succès",
                Data = await GetPatientAssuranceAsync(idPatient)
            };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Erreur lors de la mise à jour de l'assurance patient");
            return new PatientAssuranceResponse { Success = false, Message = "Erreur lors de la mise à jour" };
        }
    }

    public async Task<PatientAssuranceResponse> RemovePatientAssuranceAsync(int idPatient)
    {
        var patient = await _context.Patients.FindAsync(idPatient);
        if (patient == null)
        {
            return new PatientAssuranceResponse { Success = false, Message = "Patient non trouvé" };
        }

        patient.AssuranceId = null;
        patient.NumeroCarteAssurance = null;
        patient.DateDebutValidite = null;
        patient.DateFinValidite = null;
        patient.CouvertureAssurance = null;

        await _context.SaveChangesAsync();

        _logger.LogInformation($"Assurance retirée du patient: {idPatient}");

        return new PatientAssuranceResponse { Success = true, Message = "Assurance retirée du patient" };
    }
}
